src/Shell32.Desktop/Shell32.Helpers.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) to owners found in https://github.com/AArnott/pinvoke/blob/master/COPYRIGHT.md. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+
+namespace PInvoke
+{
+    using System;
+    using System.ComponentModel;
+    using System.Runtime.InteropServices;
+
+    /// <content>
+    /// Methods and nested types that are not strictly P/Invokes but provide
+    /// a slightly higher level of functionality to ease calling into native code.
+    /// </content>
+    public static partial class Shell32
+    {
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for GetKnownFolderPath.")]
+        public static string GetFolderPath(CSIDL csidl)
+        {
+            IntPtr pszPath;
+#pragma warning disable CS0618 // Type or member is obsolete
+            if (SHGetFolderPath(IntPtr.Zero, csidl, IntPtr.Zero, SHGetFolderPathFlags.SHGFP_TYPE_CURRENT, out pszPath) != HResult.Code.S_OK)
+#pragma warning restore CS0618 // Type or member is obsolete
+            {
+                throw new Win32Exception();
+            }
+
+            string path = Marshal.PtrToStringUni(pszPath);
+            Marshal.FreeCoTaskMem(pszPath);
+            return path;
+        }
+
+        public static string GetKnownFolderPath(Guid rfid)
+        {
+            IntPtr pszPath;
+            if (SHGetKnownFolderPath(rfid, 0, IntPtr.Zero, out pszPath) != HResult.Code.S_OK)
+            {
+                throw new Win32Exception();
+            }
+
+            string path = Marshal.PtrToStringUni(pszPath);
+            Marshal.FreeCoTaskMem(pszPath);
+            return path;
+        }
+
+        [Obsolete("As of Windows Vista, this function is merely a wrapper for GetKnownFolderID")]
+        public static unsafe string GetFolderLocation(CSIDL csidl)
+        {
+            IntPtr pidl;
+#pragma warning disable CS0618 // Type or member is obsolete
+            if (SHGetFolderLocation(IntPtr.Zero, csidl, IntPtr.Zero, 0, out pidl) != HResult.Code.S_OK)
```

Comment:
I tend to prefer accepting all success codes rather than only `S_OK`. It looks like in at least some cases the MSDN docs state S_OK is the only success code returned, but unless you've verified that for every function, can you use this syntax instead?

```csharp
if (SHGetFolderLocation(IntPtr.Zero, csidl, IntPtr.Zero, 0, out pidl).Failed)
```